### ISSUE TITLE
“Only one input size may be -1, not both 0 and 2” error in Keras

### ISSUE THREAD

**Issue by jeongHwarr** [keras-team/keras#12697]
This is summary of my model.

![2019-04-19 12;46;34](https://user-images.githubusercontent.com/16534413/56406125-cde7e580-62aa-11e9-9092-552204154038.PNG)

My model is basically similar to a convolution network.

I want my model to work regardless of the width of the input. So the width size appears as None.

and I want to attach a decoder to my model. However, when I attach the decoder, an error occurs. (If I don't attach the decoder, the program works fine)

This part is my decoder part (below)

```
if args.decoder==True:

    decoder = ConvCapsuleLayer(kernel_size=args.kernel_size, num_capsule=4, num_atoms=1, strides=1, padding='same',
                                  routings=0)(conv_cap)

    _, H, W, C, A = decoder.get_shape()

    y = layers.Input(shape=(n_class,))
    masked_by_y = Mask()([decoder, y])
    masked = Mask()(decoder)

    def shared_decoder(mask_layer):

        recon_1 = layers.Conv2DTranspose(4, (5,5), strides=(2, 2), padding='same',  kernel_initializer='he_normal', name='decoder_1', activation='relu')(mask_layer)
        recon_2 = layers.Conv2DTranspose(8, (5,5), strides=(2, 2), padding='same',  kernel_initializer='he_normal', name='decoder_2', activation='relu')(recon_1)
        recon_3 = layers.Conv2DTranspose(1, (1,1), strides=(1, 1), padding='same',  kernel_initializer='he_normal', name='decoder_3', activation='linear')(recon_2) 
        return recon_3

if args.decoder==True:
    train_model = models.Model(inputs=[x, y], outputs=[out_seg, shared_decoder(masked_by_y)]) # [x:image,y: mask] // [out_seg:length, reconstruction output]
    eval_model = models.Model(x, [out_seg, shared_decoder(masked)])
else:
    train_model = models.Model(inputs=x, outputs=out_seg)
    eval_model = models.Model(inputs=x, outputs=out_seg)
return train_model, train_model
```

`mask_1` is my Mask layer.

If a label is given, only the channel of the label is returned. `(masked_by_y)`

If a label is not given, this layer only returns channel with the largest sum of the element values ​​in `conv_capsule_layer_1`. `(masked)`

The shape of `conv_capsule_layer_1` is `(batch_size = None, height = 50, width = None, num_channel = 4, 1)`

That is, the mask layer returns a channel having the largest sum of the element values ​​among the four channels.

Then, use `Conv2DTranspose` to make it equal to the size of the original input using the returned value (output of mask layer).

However, the following error occurs.

> InvalidArgumentError (see above for traceback): Only one input size may be -1, not both 0 and 2 [[Node: mask_1/Reshape_1 = Reshape[T=DT_FLOAT, Tshape=DT_INT32, _device="/job:localhost/replica:0/task:0/device:CPU:0"](mask_1/boolean_mask/Gather, mask_1/Reshape_1/shape)]]

How can I make the length variable not using -1? I already tried that `non_zero_masked = K.reshape(non_zero,[-1, masked.shape[1], masked.shaped[2],1])`

This is function name is `call` in my `Mask` layer

```
    def call(self, inputs, **kwargs):
    if type(inputs) is list:  # true label is provided with shape = [None, n_classes], i.e. one-hot code.
        assert len(inputs) == 2
        inputs, mask = inputs
        inputs = K.squeeze(inputs, axis=-1) # [batch, input_height, input_width, num_cap, num_atom] -> [batch, input_height, input_width, num_cap] 
    else:  # if no true label, mask by the max length of capsules. Mainly used for prediction

        inputs = K.squeeze(inputs, axis=-1) #[batch, input_height, input_width, num_cap]
        x = K.softmax(K.sqrt(K.sum(K.square(inputs), axis=(1,2)) + K.epsilon())) # x: [batch, 4]
        mask = K.one_hot(indices=K.argmax(x, 1), num_classes=x.get_shape().as_list()[1]) # mask: [batch,4]


    expand_mask = K.reshape(mask,[-1,1,1,mask.shape[1]]) #[batch_size, 1, 1, num_class]
    masked = inputs*expand_mask
    non_zero = tf.boolean_mask(masked, tf.not_equal(masked,0))
    non_zero_masked = K.reshape(non_zero,[-1, masked.shape[1], -1,1])
    return non_zero_masked
```

How can I solve it?

**Comment by jvishnuvardhan** [comment]
Please fill the issue [template](https://github.com/keras-team/keras/issues/new?template=a--tensorflow-backend-users.md). Could you update them if they are relevant in your case, or leave them as N/A? Along with the template, please provide as many details as possible to find the root cause of the issue. It would be great if you can provide a small code to reproduce the error. Thanks!
